dds/DCPS/NetworkResource.cpp

```diff
@@ -794,18 +826,27 @@ ACE_INET_Addr choose_single_coherent_address(const String& address, bool prefer_
     ACE_INET_Addr temp;
     temp.set_addr(&addr, sizeof addr);
     temp.set_port_number(port_number, 1 /*encode*/);
-    addresses.push_back(temp);
+    for (size_t i = 0; i < addr_count; ++i) {
+      if (temp == addr_array[i]) {
+        addresses.push_back(temp);
 #ifdef ACE_WIN32
-    if (it != addr_cache_map_.end()) {
-      it->second.second.insert(temp);
-    }
+        if (it != addr_cache_map_.end()) {
+          it->second.second.insert(temp);
+        } else {
+          OPENDDS_SET(ACE_INET_Addr) my_set;
+          my_set.insert(temp);
+          addr_cache_map_[host_name] = make_pair(now, my_set);
+        }
 #endif /* ACE_WIN32 */
+      }
+    }
   }
 
 #ifdef ACE_WIN32
   g.release();
 #endif /* ACE_WIN32 */
 
+  delete[] addr_array;
```

Comment:
I don't think we need to do this here (see other comment), but if we did I'd prefer to see something like ArrayGuard above to handle the deletion. I suppose the same goes for freeaddrinfo. Scoped protection object is better in case we get an exception or if we add an early return for some reason in the future.